### ISSUE TITLE
Add a variable to kepler_geo.py for current station location

### ISSUE THREAD

**Issue by hcarter333** [hcarter333/rm-rbn-history#9]
Make the KD0FNR station location a variable so it can be easily reassigned when the station is remote.

**Comment by hcarter333** [comment]
Also add a time zone adjust variable.

**Comment by hcarter333** [comment]
The initial solution of simply adding a new time zone and location that will be picked up for future RBN entries didn't work. It munged incorrect locations into several previously handled entries because the query returns up to 100 rows. Some of those rows had already been handled, but then, when they were handled 'again' they got the new location and time, and that caused them to be 'different' and therefore added to the presentation report. I'm rolling the presentation report back to it's previous version.

**Comment by hcarter333** [comment]
The next proposal would be to patch the lines that are picked up in the action after the diff of the raw data. In other words, the raw data will always have San Francisco and -8 as location and time. The action will switch out the location and time differential for the new lines that come in. It's harder, but it'll do the trick.

**Comment by hcarter333** [comment]
Maybe an easier trick is to not append each line if the first field, they key, is already in the presentation file.

![image](https://user-images.githubusercontent.com/363004/215237807-fb057111-0160-4a7f-9da0-0ed6768fd55a.png)

**Comment by hcarter333** [comment]
A python script will do the following:
1. open the existing rm_rnb_history_pres.csv
2. Assemble the keys into a list
3. As each line arrives on stdin if the key is in the list do nothing. If it is not, print it to standard out.

Tests:
1. Make a file with 3 lines from rm_rnb_history_pres.csv. Check that the keys are reported correctly in the list returned from make_key_list in new_key_cat.py
2. Make a copy of the 3 line file. Add two lines that use keys not already in the file.
3. Check that the 2 new lines are the only ones returned.




Post on how to pull the first field of a csv file into a list
https://stackoverflow.com/questions/39662891/read-in-the-first-column-of-a-csv-in-python

**Comment by hcarter333** [comment]
Took a while, but got the first test case passing!!!
![image](https://user-images.githubusercontent.com/363004/215251673-acbd7500-b071-4cd1-87fc-fcfdbf28e57e.png)

**Comment by hcarter333** [comment]
So, this has become a bit of an oddysey, but here's more cool stuff:
The output python script new_key_cat.py is supposed to present only lines of output (to stdout) that do not contain keys that already exist in rm_rnb_history_pres.csv. Tests are being built for the script using pytest. pytest mutes stdin because the tests are intended to be automated rather than interactive. This makes sense to me, so I'm changing the method in the script that handles input to look for a file name argument, and if it doesn't have one to use stdin.

The second issue though was how to write an automated test case that looks at stdout. It turns out pytest has that covered. It's all written up at
https://docs.pytest.org/en/7.1.x/how-to/capture-stdout-stderr.html

And here's the pertinent code example from that page:
<pre>
def test_myoutput(capsys):  # or use "capfd" for fd-level
    print("hello")
    sys.stderr.write("world\n")
    captured = capsys.readouterr()
    assert captured.out == "hello\n"
    assert captured.err == "world\n"
    print("next")
    captured = capsys.readouterr()
    assert captured.out == "next\n"
</pre>

The test case for new_key_cat specified above will send in a file with two keys that already exist in rm_rnb_history_pres.csv. The test case will check to make sure they are not output by doing the following
<pre>
def test_new_key_cat_2(capsys):  # or use "capfd" for fd-level
    new_key_cat("testfile_with_existing_keys")
    captured = capsys.readouterr()
    assert not captured.out.find("existing_key_1")
</pre>

So, if an existing key does turn up in the output, the test case will fail. A more exhaustive test would be to look for all the existing keys as in:
<pre>
def test_new_key_cat_2(capsys):  # or use "capfd" for fd-level
    new_key_cat("testfile_with_existing_keys")
    captured = capsys.readouterr()
    for entry in make_key_list("rm_rnb_history_pres.csv"):
        assert not captured.out.find(entry)
</pre>

The documentation for capsys offers far more detail:
https://docs.pytest.org/en/7.1.x/reference/reference.html?highlight=capsys#std-fixture-capsys

I'm kinda falling in love with pytest.

**Comment by hcarter333** [comment]
OK, the code has been added. The first thing to do, I suppose, will be to see if pytest really can work over here. I think this should do the trick:
<pre>
- name: Test with pytest
  run: |-
    pip install pytest
    pip install pytest-cov
    pytest 
</pre>
And that in fact completely did the trick!
![image](https://user-images.githubusercontent.com/363004/215346256-7b533bea-153e-4fb0-85bd-3e8aacbe7f80.png)

**Comment by hcarter333** [comment]
The next thing to do is to test out the script in real life. Here's the flow:
1. Add the unique key filter new_key_cat.py to the workflow like so:
<pre>
git diff HEAD rm_rnb_history.csv | cat | egrep "^\+[^\+]" | cut -c2- | new_key_cat.py >> rm_rnb_history_pres.csv
</pre>
2. commit the workflow. There should be no changes to rm_rnb_history_pres.csv because there's on new data.
3. Change the variables in geo_kepler.py to point at Three Rivers.
4. commit geo_kepler.py. There should be no changes to rm_rnb_history_pres.csv because no new keys will be generated by the workflow.
5. Remove the lines starting with 1/25/23 from the presenation database rm_rnb_history_pres.csv (This leaves us in a position where the incoming keys from geo_kepler.py will be unique.)
6. Commit the modified file.
**What should happen:**
The automated update should pick up all the rows from then till now; they should get Three Rivers' location; they should be added to rm_rnb_history_pres.csv.
**Repeat the process for the Organ Mountain**
Repeat the above for Organ Mountain, (change the geo_kepler.py variables; delete the rows in rm_rnb_history_pres.csv from 1/25 late afternoon on.)

**Comment by hcarter333** [comment]
github isn't finding new_cat_key.py.
![image](https://user-images.githubusercontent.com/363004/215347021-f6b8e5d3-69ab-4ebd-891b-63630834ea8a.png)
It is finding kepler_geo.py, so what's the difference?
It's that python is used to execute kepler_geo.py as in 
<pre>
python kepler_geo.py
</pre>
I'll try that
Using
<pre>
git diff HEAD rm_rnb_history.csv | cat | egrep "^\+[^\+]" | cut -c2- | python new_key_cat.py >> rm_rnb_history_pres.csv
</pre>
Did the trick:
![image](https://user-images.githubusercontent.com/363004/215347420-1672a6a5-e7bb-4f17-a12c-38850df57831.png)

**Comment by hcarter333** [comment]
As specified, rm_rnb_history_pres.csv was not changed

![image](https://user-images.githubusercontent.com/363004/215347505-ea31aaa2-de89-4e3f-9f44-dec447371449.png)

Now, change the variables in kepler_geo.py to point to Three Rivers:
<pre>
#San Francisco
geo_station = "-122.42299,37.72286"
time_diff = 8

#Three Rivers Petroglyph Site
#geo_station = "-106.006336, 33.346341"
#time_diff = 7

</pre>

Goes to:

<pre>
#San Francisco
#geo_station = "-122.42299,37.72286"
#time_diff = 8

#Three Rivers Petroglyph Site
geo_station = "-106.006336, 33.346341"
time_diff = 7

</pre>

No new unique keys, so, sure enough, no changes to rm_rnb_history_pres.csv:
![image](https://user-images.githubusercontent.com/363004/215347676-312f1a22-c052-4f59-a43a-137ad9e88c8e.png)


The next step is to cut the rows from 1/25/23 on from rm_rnb_history_pres.csv
There were 95 entries. We're just going to pull this off while fitting in the 100 max entries output by RBN.
And that did not work. There were no changes to rm_rnb_history_pres.csv.

If I look at the rm_rnb_history.csv, there were definitely changes. [For example:](https://github.com/hcarter333/rm-rbn-history/commit/8ef4d48202f2250a0bfd63f97ea33d0bcc9882e7#r98508973)
![image](https://user-images.githubusercontent.com/363004/215348224-29c7884e-9716-4262-8dc5-85ef8e653d9b.png)
The question becomes, is that key in rm_rnb_history_pres.csv?

It is not.
Then, the question becomes, did new_cat_key.py output anything?

I may have missed something. Deleting the rows from rm_rnb_history.csv to see if this will just rectify itself.
Nope. The changes still weren't propagated:
![image](https://user-images.githubusercontent.com/363004/215348447-c1d10349-547b-4829-bdb9-71bf642daf5d.png)

So, that leaves there not being any output from new_key_cat.py.

Let's check:
Oh! Oh geesh! There's no script to run in new_key_cat.py. Only two methods.
OK, that's fixed now. There's a script nkc.py that uses the method.

**Comment by hcarter333** [comment]
Furthermore, at some point, the lines wound up back in rm_rnb_history_pres.csv?
Perhaps just locally? Nope. It happened right after I removed them and committed. They were no longer unique keys, so, they were added back in before I changed the kepler_geo location variables, and then rejected any changes after that. So, if I remove them one more time, we should be off and running.
Argggghhhhh! Actually, they are gone, they just hadn't managed to make it in on the git pull command locally.
So, back to debugging locally.
OK, it looks like nkc.py works! The whole issue was resolved above. I'm changing the workflow to use nkc.py.

Still nothing. So, now it's working locally, but not in the repository. Output the script to stderr as well to see what's happening.

Andddd!!!! The problem is that the entries in rm_rnb_history, after several iterations, have already been updated with the new location, so the diff on that file isn't picking up any changes to send to nkc.py.
![image](https://user-images.githubusercontent.com/363004/215350825-6a4b6af9-3e57-4f0e-a60f-017890bcc483.png)


Let's try removing the entries there again.
Success?
![image](https://user-images.githubusercontent.com/363004/215350954-c9681a26-b451-4a2e-a29b-c06dacd7eb0b.png)
Success!!!
![image](https://user-images.githubusercontent.com/363004/215351028-53ae2f9e-e367-4405-866b-16542bfe8837.png)

Now to update for the Organs by:
1. Change kepler_geo.py geo values
2. Remove the Organ lines from rm_rbn_history.py
3. Remove the lines from rm_rbn_history_pres.py
4. Repeat step 2 to get things to fully propagate

Question: with unique key checking, do we still need the git diff filter?
